### ISSUE TITLE
Assertion failed: "Ptr must be a pointer to Val type!"

### ISSUE THREAD

**Issue by rozaliev** [rust-lang/rust#17257]
rust version: rustc 0.12.0-pre-nightly (09abbbdaf 2014-09-11 00:05:41 +0000)
```
Assertion failed: (getOperand(0)->getType() == cast<PointerType>(getOperand(1)->getType())->getElementType() && "Ptr must be a pointer to Val type!"), function AssertOK, file /Users/rustbuild/src/rust-buildbot/slave/nightly-mac/build/src/llvm/lib/IR/Instructions.cpp, line 1085.
```

code to reproduce:
```
trait Tr: Clone {
  fn clone_to_tobj(&self) -> Box<Tr> {
    return box self.clone()
  }
}

#[deriving(Clone)]
struct A;
impl Tr for A {
}


fn main() {
   let a: Box<Tr> = box A;
}
```

but this compiles fine
```
trait Tr: Clone {
  fn clone_to_tobj(&self) -> Box<Tr> {
    box self.clone()
  }
}

#[deriving(Clone)]
struct A;
impl Tr for A {
}


fn main() {
   let a: Box<Tr> = box A;
}
```

**Comment by JustAPerson** [comment]
Updating for current Rust (on playpen):
**tl;dr**: *only @P1start example is still broken* (See below for updated syntax).

I was unsure how to update @rozaliev's example because of object safety errors.

@pnathan's example *compiles fine*:
```rust
fn read_file(path : String) -> Box<[u8]> {
    return Box::new([])
}


fn main() {
}
```

@mdup's example *compiles fine*:
```rust
trait Vehicle { fn stuff(&self) {} }
struct Car;
impl Vehicle for Car { }

fn build_car<'a>() -> Box<Vehicle + 'a> {
    loop {
        return Box::new(Car);
    }
}

fn main() { }
```

@tomjakubowski's example *compiles fine* as is.

@P1start's example is *still broken* and causes a coredump:
```rust
struct Foo<T: ?Sized> {
    x: u8,
    y: T,
}

fn main() {
    let foo: &Foo<[isize]> = &Foo { x: 1, y: [1, 2, 3] };
    match *foo {
        Foo { x, ref y } => {},
    }
}
```

@tomassedovic's example *compiles fine* even with `rustc -C debuginfo=2`
```rust
#![feature(unboxed_closures)]

fn main() {
    let i: isize = 0;
    let some_closure = || {
        let _capture = i;
    };
}
```

All of @hugwijst's variants now *compile fine*.

**Comment by excaliburHisSheath** [comment]
I ran into this issue on the latest nightly with the following example:

    type EngineInit = extern fn (Box<Window>) -> Box<Engine>;
    type EngineUpdateAndRender = extern fn (&mut Engine);

    fn main() {
        // Open the game as a dynamic library;
        let lib = DynamicLibrary::open(Some(Path::new("gunship-24517baeade73325.dll"))).unwrap();

        let engine_init = unsafe {
            mem::transmute::<*mut EngineInit, EngineInit>(lib.symbol("engine_init").unwrap())
        };

        let engine_update_and_render = unsafe {
            mem::transmute::<*mut EngineUpdateAndRender, EngineUpdateAndRender>(lib.symbol("engine_update_and_render").unwrap())
        };

        let mut window = Window::new("Game Window");
        let mut engine = engine_init(window);
        engine_update_and_render(&mut engine);
    }

If I change the first line to

    type EngineInit = extern fn (Box<Window>) -> Engine;

It compiles fine.

The error I get:

![](http://i.imgur.com/tB83w6Y.png)

**Comment by pmarcelll** [comment]
@JustAPerson, I couldn't trigger an ICE with any of the code it this issue, but I found something interesting: as you said, this compiles fine:
```rust
fn read_file(path : String) -> Box<[u8]> {
    return Box::new([])
}

fn main() {
}
```
But I think, it shouldn't. I've never seen return without a semicolon mentioned as valid Rust code.

**Comment by pmarcelll** [comment]
Ok, it turned out to be valid syntax (https://github.com/rust-lang/rust/issues/26425) and I still can't reproduce the ICE :+1:

**Comment by alexcrichton** [comment]
Thanks @pmarcelll!
